services/venice-router/src/main/java/com/linkedin/venice/router/api/path/VeniceMultiKeyPath.java

```diff
@@ -230,14 +240,18 @@ public int getLongTailRetryMaxRouteForMultiKeyReq() {
 
   @Override
   public boolean isLongTailRetryAllowedForNewRoute() {
+    // We are keeping both features for now. Retry is only allowed if both conditions pass:
+    // 1. retry attempts < longTailRetryMaxRouteForMultiKeyReq (if set).
+    // 2. we still have retry budget according to retry manager.
     if (longTailRetryMaxRouteForMultiKeyReq == -1) {
-      // feature is disabled
-      return true;
-    }
-    if (longTailRetryMaxRouteForMultiKeyReq <= 0) {
+      // max route feature is disabled
+      return retryManager.isRetryAllowed();
+    } else if (longTailRetryMaxRouteForMultiKeyReq <= 0) {
       return false;
+    } else {
+      return (currentAllowedRetryRouteCnt.incrementAndGet() <= longTailRetryMaxRouteForMultiKeyReq)
```

Comment:
This is slightly different from my understanding..

Let us say the original request contains 1000 keys and it is equally divided into 10 routes, so each route would take 100 keys.
If Router retries one slow route, which contains 100 keys, based on the current mechanism, the retry will be counted as `1`. If the route retry fanout size is 10, I think we should count it as `10` instead of `1`, as the scattered requests are the ones being sent to the storage node, and retry storm is mainly about the small requests sent to the storage nodes.

And this logic can be put in `VeniceDelegateMode`, where we would know the fanout size for both original request and the retry request for each Route.
Let me know if this is not clear.